// adapt-schemas/lib/XSSDefaults.js
/**
 * Default XSS whitelist for HTML tags and attributes
 */
export default {
  a: ['class', 'href', 'rel', 'target', 'title'],
  abbr: ['title'],
  address: [],
  area: ['alt', 'coords', 'href', 'shape'],
  article: [],
  aside: ['aria-hidden', 'class', 'role'],
  audio: [
    'autoplay',
    'controls',
    'crossorigin',
    'loop',
    'muted',
    'preload',
    'src'
  ],
  b: [],
  bdi: ['dir'],
  bdo: ['dir'],
  big: [],
  blockquote: ['cite'],
  br: [],
  button: ['class'],
  caption: [],
  center: [],
  cite: [],
  code: [],
  col: ['align', 'span', 'valign', 'width'],
  colgroup: ['align', 'span', 'valign', 'width'],
  data: [],
  dd: [],
  del: ['datetime'],
  dfn: [],
  details: ['open'],
  div: [
    'aria-describedby',
    'aria-description',
    'aria-label',
    'aria-hidden',
    'class',
    'role',
    'tabindex'
  ],
  dl: [],
  dt: [],
  em: [],
  figcaption: [],
  figure: ['class'],
  font: ['color', 'face', 'size'],
  footer: [],
  h1: ['class'],
  h2: ['class'],
  h3: ['class'],
  h4: ['class'],
  h5: ['class'],
  h6: ['class'],
  header: [],
  hr: [],
  i: [],
  img: [
    'alt',
    'aria-hidden',
    'aria-label',
    'class',
    'height',
    'loading',
    'src',
    'title',
    'width'
  ],
  ins: ['datetime'],
  kbd: [],
  li: ['class'],
  mark: [],
  math: [],
  mfrac: [],
  mi: [],
  mn: [],
  mo: [],
  mover: [],
  mrow: [],
  ms: [],
  mspace: [],
  msub: [],
  msubsup: [],
  msup: [],
  mtext: [],
  munder: [],
  munderover: [],
  nav: [],
  ol: ['class'],
  p: ['lang'],
  pre: [],
  q: [],
  rp: [],
  rt: [],
  ruby: [],
  s: [],
  samp: [],
  section: [],
  small: [],
  span: [
    'aria-describedby',
    'aria-description',
    'aria-label',
    'aria-hidden',
    'class',
    'role',
    'tabindex'
  ],
  sub: [],
  summary: [],
  sup: [],
  strong: [],
  strike: [],
  table: ['align', 'border', 'width', 'valign'],
  tbody: ['align', 'valign'],
  td: ['align', 'colspan', 'rowspan', 'valign', 'width'],
  tfoot: ['align', 'valign'],
  th: ['align', 'colspan', 'rowspan', 'valign', 'width'],
  thead: ['align', 'valign'],
  time: [],
  tr: ['align', 'rowspan', 'valign'],
  tt: [],
  u: [],
  ul: ['class'],
  var: [],
  video: [
    'autoplay',
    'controls',
    'crossorigin',
    'loop',
    'muted',
    'playsinline',
    'poster',
    'preload',
    'src',
    'height',
    'width'
  ],
  wbr: []
}
